pkg/chains/formats/intotoite6/intotoite6.go

```diff
@@ -0,0 +1,351 @@
+package intotoite6
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/tektoncd/chains/pkg/chains/formats"
+	"github.com/tektoncd/chains/pkg/config"
+
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+
+	"github.com/google/go-containerregistry/pkg/name"
+)
+
+const (
+	tektonID = "https://tekton.dev/attestations/chains@v1"
+)
+
+var ErrNoBuilderID = errors.New("no builder id configured")
+
+const (
+	commitParam        = "CHAINS-GIT_COMMIT"
+	urlParam           = "CHAINS-GIT_URL"
+	ociDigestResult    = "IMAGE_DIGEST"
+	chainsDigestSuffix = "_DIGEST"
+)
+
+type artifactType int
+
+const (
+	artifactTypeUnknown artifactType = iota
+	artifactTypeOCI
+)
+
+type InTotoIte6 struct {
+	builderID string
+}
+
+type artifactResult struct {
+	ParamName  string
+	ResultName string
+	Type       artifactType
+}
+
+type vcsInfo struct {
+	Commit string
+	URL    string
+}
+
+func NewFormater(cfg config.Config) (formats.Payloader, error) {
+	if cfg.Builder.ID == "" {
+		return nil, ErrNoBuilderID
+	}
+
+	return &InTotoIte6{builderID: cfg.Builder.ID}, nil
+}
+
+func (i *InTotoIte6) CreatePayload(obj interface{}) (interface{}, error) {
+	var tr *v1beta1.TaskRun
+
+	switch v := obj.(type) {
+	case *v1beta1.TaskRun:
+		tr = v
+	default:
+		return nil, fmt.Errorf("unsupported type: %s", v)
+
+	}
+
+	if tr.Spec.TaskRef == nil {
+		return nil, fmt.Errorf("provided TaskRef is nil for %s", tr.Status.PodName)
+	}
+
+	// Here we translate a Tekton TaskRun into an InToto ite6 attestation.
+	// At a high leevel, the  mapping looks roughly like:
+	// Podname -> Builder.Id
+	// Results with name *_DIGEST -> Subject
+	// Step containers -> Materials
+	// Params with name CHAINS-GIT_* -> Materials and recipe.materials
+	// tekton-chains -> Recipe.type
+	// Taskname -> Recipe.entry_point
+	att := in_toto.ProvenanceStatement{
+		StatementHeader: in_toto.StatementHeader{
+			Type:          in_toto.StatementInTotoV01,
+			PredicateType: in_toto.PredicateProvenanceV01,
+		},
+		Predicate: in_toto.ProvenancePredicate{
+			Builder: in_toto.ProvenanceBuilder{
+				ID: i.builderID,
+			},
+			Recipe: in_toto.ProvenanceRecipe{
+				Type:       tektonID,
+				EntryPoint: tr.Spec.TaskRef.Name,
```

Comment:
Do you think it would be useful to include `Arguments` and `Environment` to the recipe as well? I'm not sure if we need people to be able to rebuild the TaskRun from the annotation, but we'd probably want them in there if that's the goal.

Reply:
Great comment. For `Environment` I think we should skip it now, as by the in-toto spec states that any included environment value should be _only_ the part of the "build API", and from Chain's perspective, it's hard to understand that in general (such semantic would be defined in the tasks which I don't know of a good and reliable way for Chains to interpret).

For arguments it's a slightly similar story, what goes in the `Arguments` value is "external input that influenced the build", and better would be to rely on the `Entrypoint` alone (which the spec also recommends). 
The problem with the PR in the current shape as the value for the `Entrypoint` is not fully correct, currently it sets the value to the TaskName.

So with that in mind, my wish would be to have an annotation or some type hinting that would allow the task to specify the _real_ arguments to the build, and the `Recipe.Type`. As an example, we have a `maven` task, that accepts parameter `GOALS`, where `GOALS` might be `[install, -DskipTests]`. And with that we should produce an attestation with `Recipe.Type: maven`, `Entrypoint`: `install -DskipTests` or similar, but I don't have spent too much time thinking of such a solution yet. But I will soon :)

Reply:
SGTM, thanks for explaining!